### PR TITLE
Implementing transpose and view methods for PackedTensorAccessor

### LINKED ISSUE #60431: Implement simple view methods for TensorAccessor/PackedTensorAccessor
THCDeviceTensor (the old version of PackedTensorAccessor) supports a bunch of handy utilities like transpose and view (and some more exotic ones like downcastInner). PackedTensorAccessor, the modern replacement, doesn't support any of these. They should!

### PR DESCRIPTION
**Pull request by ramvenkat98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60968**

See https://github.com/pytorch/pytorch/issues/60431

The methods are implemented to match the functionality of THCDeviceTensor, which was used for a similar purpose as TensorAccessor/PackedTensorAccessor.

Corner cases are:
1. For the 1d case, transpose returns a PackedTensorAccessor that is an identical copy of the original; this mirrors the behavior of THCDeviceTensor
2. The 1d case for view is not implemented - the reason for this is that in THCDeviceTensor, there would be no 1d case for view - it would be forbidden by static asserts. So this matches that behavior.

Differential Revision: [D29505537](https://our.internmc.facebook.com/intern/diff/D29505537)

### CODE REVIEW

File: aten/src/ATen/core/TensorAccessor.h [hunk 1]
```diff
@@ -187,6 +196,39 @@ class GenericPackedTensorAccessor : public GenericPackedTensorAccessorBase<T,N,P
...
+    this->bounds_check_(dim2);
+    GenericPackedTensorAccessor<T, N, PtrTraits, index_t> result(
+        this->data_, this->sizes_, this->strides_);
+    std::swap(result.strides_[dim1], result.strides_[dim2]);
+    std::swap(result.sizes_[dim1], result.sizes_[dim2]);
+    return result;
+  }
+
+  /// Returns a PackedTensorAccessor that is a view of the `SubDim`-dimensional
+  /// slice of this tensor, starting at `at`
```
Comment (ezyang) [inline aten/src/ATen/core/TensorAccessor.h:217]: This doesn't match the signature of Tensor.view?

File: aten/src/ATen/test/packedtensoraccessor_test.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  Tensor t = rand(sizes, CPU(kFloat));
+  auto original = t.packed_accessor64<float, dimension, DefaultPtrTraits>();
+  auto transposed = original.transpose(0, 2);
+  ASSERT_EQ(original.size(0), transposed.size(2));
+  ASSERT_EQ(original.size(1), transposed.size(1));
+  ASSERT_EQ(original.size(2), transposed.size(0));
+  for (int i = 0; i < sizes[0]; i++)
+    for (int j = 0; j < sizes[1]; j++)
+      for (int k = 0; k < sizes[2]; k++)
+        ASSERT_EQ(original[i][j][k], transposed[k][j][i]);
```
Comment (ezyang) [inline aten/src/ATen/test/packedtensoraccessor_test.cpp:27]: nit: use braces

**Comment by ezyang** [comment]
We're gonna drop view() and this is good to go.

@ramvenkat98 Do you know if it is easy or not to make TensorAccessor also support transpose, but by returning a PackedTensorAccessor? If we gotta copy paste code to make it work, not worth it, we'll just do it as is here.

**Comment by ramvenkat98** [comment]
Copy-pasting the code would work in letting transpose run (on hosts), but I'm not aware of a cleaner way of doing this.

I guess that in general, as more functions are added to this file, there might be a decent amount of code duplication to deal with (since the implementations have to be written for both the base case and the recursive case, for both TensorAccessor and PackedTensorAccessor).

As a random thought, one way of getting around this is might be to make TensorAccessor a base class and PackedTensorAccessor inherit from it but with PackedTensorAccessor having two extra data members - the size and stride array. In this case, it would be possible to implement most methods directly in TensorAccessor (like transpose - which will always return a PackedTensorAccessor - and I think even the existing [] operators - which always return a TensorAccessor - would not need to be duplicated).

There might be some extra difficulties to this that I'm missing (like the host/device stuff), and it adds one extra level of inheritance since now there's PackedTensorAccessor <- TensorAccessor <- TensorAccessor base but it can help get rid of PackedTensorAccessorBase and perhaps avoid some duplication...

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Approved. For the hypothetical other improvements, another way to do it without duplication and without inheritance is to set up a traits class specifying what operations PackedTensorAccessor and TensorAccessor support, and then implement transpose generically over the trait.

**Comment by ramvenkat98** [comment]
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29459384).

**Comment by ramvenkat98** [comment]
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29459384).

**Comment by driazati** [comment]
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29503741).

**Comment by ramvenkat98** [comment]
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29505537).

**Comment by ramvenkat98** [comment]
Closed: see https://github.com/pytorch/pytorch/pull/61114 instead

**Pull request closed without merging**